src/classes/ALLO_AllocationsDMLService.cls

```diff
@@ -0,0 +1,413 @@
+        private set;
+    }
+
+    /** @description Makes sure recErrs is initialized, if one was not passed using the default Allocation
+     *               Context Type */
+    private ERR_RecordError getErrorLogger() {
+        if (this.recErrs == null) {
+            this.recErrs = new ERR_RecordError(this.contextType);
+        }
+
+        return this.recErrs;
+    }
+
+    /**
+     * @description provides a non default record errors instance to use when logging errors
+     * @param altRecErrs Already initialized instance of the ERR_RecordError class
+     * @return Class Instance
+     */
+    public ALLO_AllocationsDMLService withAlternateRecordErrorsLogger(ERR_RecordError altRecErrs) {
+        this.recErrs = (altRecErrs != null) ? altRecErrs : this.recErrs;
+        return this;
+    }
+
+    /**
+     * @description Provides an allocations wrapper to the class for processing
+     * @param allocationsWrapper Allocation Wrapper with DML to be processed
+     * @return Class Instance
+     */
+    public ALLO_AllocationsDMLService withAllocationsWrapper(ALLO_AllocationsWrapper allocationsWrapper) {
+        this.allocationsWrapper = allocationsWrapper;
+        return this;
+    }
+
+    /**
+     * @description Passes a Set of Processes that should be excluded if errors occur
+     * @param processExclusions the Set of processDefinitionTypes to be added as record exclusions if there are
+     *        DML Errors
+     * @return Class Instance
+     */
+    public ALLO_AllocationsDMLService withProcessExclusions(Set<ALLO_AllocationsService.processDefinitionType> processExclusions) {
+        this.processExclusions = processExclusions;
+        return this;
+    }
+
+    /**
+     * @description Processes the DML for the Allocation Results and updates the Allocations Wrapper with the results,
+     *              adding exclusions and errors as needed
+     **/
+    public void processAllocationResults() {
+        Boolean dmlErrors = false;
+        Integer dmlIteration = 0;
+
+        // Clone existing allocation lists in case they need to be restored when rolling back
+        Map<Id, List<Allocation__c>> backupOppsAllocations;
+        Map<Id, List<Allocation__c>> backupPaymentAllocations;
+
+        Savepoint sp = Database.setSavepoint();
+
+        do {
+            // Clone existing allocation lists in case they need to be restored when rolling back
+            backupOppsAllocations = allocationsWrapper.oppsAllocations.clone();
+            backupPaymentAllocations = allocationsWrapper.paymentAllocations.clone();
+
+            dmlIteration++;
+            dmlErrors = false;
+
+            // Process Deletes
+            dmlErrors = (dmlErrors || processAllocationDeleteResults());
+
+            // Process Inserts
+            dmlErrors = (dmlErrors || processAllocationInsertResults());
+
+            // Process Updates
+            dmlErrors = (dmlErrors || processAllocationUpdateResults());
+
+            if (dmlErrors) {
```

Comment:
Could this be
```
if (!dmlErrors) {
   continue;
}
````

Reply:
No.  It needs to go through all the dml types before it rollback and filters the results.  If I had an unexpected error in the insert step, and skipped the remainder, and filtered, on the second try I could still hit errors on the update or delete steps.  I have to do all three and then rollback, filter, and retry.

Reply:
Yes. Agree that it does need to go through the 3 above, but the large if block below (line 181 through 196) that starts with `if (dmlErrors) {` could be outdented if no errors. Just helps with readability.

Reply:
@force2b - I pushed the items in the if block into their own method.  That should help readability

Reply:
Even better.